app/controllers/oai_controller.rb

```diff
@@ -93,23 +108,62 @@ def identify_params
-  def list_record_normalized_dump_candidates
-    @list_record_normalized_dump_candidates ||= begin
-      current_full_dump = @stream.current_full_dump
+  # Get a page of OAI-XML records and a token pointing to the next page
+  def next_record_page(token = nil)
+    # parse the token if we were provided one
+    set, page, from_date, until_date = OaiConcern::ResumptionToken.decode(token) if token
+    page = page.to_i
+
+    # filter normalized dumps and get the corresponding OAI-XML pages
+    # NOTE: each page is guaranteed to have < OAIPMHWriter::max_records_per_file,
+    # but some pages will have exactly that number and others won't. The sequence
+    # isn't predictable; the only thing we promise is that all pages are less than
+    # that size.
+    pages = normalized_dumps(set, from_date, until_date).flat_map(&:oai_xml_attachments)
+    raise OaiConcern::NoRecordsMatch if pages.empty?
+
+    # generate a token for the next page, if there is one
+    token = case page
+            when (0...pages.count - 1)
+              OaiConcern::ResumptionToken.encode(set: set, page: page + 1, from_date: from_date, until_date: until_date)
+            when (pages.count - 1)
+              nil
+            else
+              raise OaiConcern::BadResumptionToken
+            end
 
-      dumps = @stream.normalized_dumps.where(id: current_full_dump.id).or(current_full_dump.deltas).order(:created_at)
+    # return the current page and the token for the next page, if any
+    [pages[page], token]
+  end
 
-      if list_records_params[:from]
-        dumps = dumps.where(created_at: (Time.zone.parse(list_records_params[:from]).beginning_of_day)...)
-      end
-      dumps = dumps.where(created_at: ...(Time.zone.parse(list_records_params[:until]).end_of_day)) if list_records_params[:until]
-      dumps = dumps.where(id: list_records_params[:resumptionToken]...) if list_records_params[:resumptionToken]
+  # Get NormalizedDumps from a particular stream or created between two dates
+  # NOTE: this likely needs work to memoize/tune, but it should be called
+  # repeatedly by next_record_page with the same arguments
+  def normalized_dumps(set, from_date, until_date)
+    # get stream by id, or all default streams if not specified
+    streams = if set.present?
+                Stream.where(id: set)
+              else
+                Stream.joins(:default_stream_histories).joins(normalized_dumps: :oai_xml_attachments).distinct
+              end
+
+    # get all dumps in this stream, optionally between two dates; error if none
+    dumps = filter_dumps(streams, from_date, until_date)
+    raise OaiConcern::NoRecordsMatch if dumps.empty?
+
+    dumps
+  end
 
-      dumps
-    end
+  def filter_dumps(streams, from_date, until_date)
+    # get candidate dumps (the current full dump and its deltas for each stream)
+    dumps = streams.flat_map(&:current_dumps).sort_by(&:created_at)
+
+    # filter candidate dumps (by from date and until date)
```

Comment:
Could these `select`s end up as part of the scoping for current_dumps?

Reply:
Would that just involve adding optional arguments to `current_dumps`, like:
```rb
  def current_dumps(from_date, until_date)
...
  end
```

If so, I'm not sure how to call it in a way such that I can still use ActiveRecord `select`, since `current_dumps` on its own just returns an array. I think I need a little more explanation to implement it, but I'm on board with it.

Reply:
@thatbudakguy I think `current_dumps` could be re-written using ActiveRecord. Maybe there's a more clever way to do this, but one way would be to first get the id of the most recent full dump and then write a query with `or` to fetch the full dump and its deltas. You can also build  up a query with conditional clauses (such as for filtering by from/until dates, and maybe to support paging too).:

```ruby
full_dump = normalized_dumps.full_dumps.order(created_at: :desc).limit(1)
full_dump_id = full_dump.first.id
dumps_query = NormalizedDump.where(id: full_dump_id).or(NormalizedDump.where(full_dump_id: full_dump_id)).order(created_at: :asc)
if from_date && until_date
  dumps_query = dumps_query.where('created_at BETWEEN ? AND ?', Time.zone.parse(from_date).beginning_of_day, Time.zone.parse(until_date).end_of_day)
else
.... etc.
end
dumps_query
```

I think the paging could be handled here too, by applying `limit(1)` and `offset(page)` to the query.

Reply:
@thatbudakguy I was messing around with this also and reaching similar conclusions.

Reply:
I think filtering the ActiveStorage::Attachment objects may be feasible if we start with the list of relevant NormalizedDump ids: `record_id` in the query:

```
ActiveStorage::Attachment.where(record_type: 'NormalizedDump', name: 'oai_xml', record_id: [60,62]).order(crea
ted_at: :asc).limit(1).offset(2)
```